<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>rCOGS Quickstart • rCOGS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="rCOGS Quickstart">
<meta property="og:description" content="rCOGS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rCOGS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Quickstart.html">rCOGS Quickstart</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Quickstart_files/header-attrs-2.3/header-attrs.js"></script><script src="Quickstart_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>rCOGS Quickstart</h1>
                        <h4 class="author">Olly Burren</h4>
            
            <h4 class="date">2020-07-31</h4>
      
      
      <div class="hidden name"><code>Quickstart.Rmd</code></div>

    </div>

    
    
<div id="what-is-rcogs" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-rcogs" class="anchor"></a>What is rCOGS ?</h2>
<p><strong>rCOGS</strong> is a package for integrating GWAS summary statistics with the <strong>p</strong> romoter <strong>c</strong> apture Hi-C (pcHi-C) contact maps in order to prioritise genes for functional followup. COGS stands for <strong>C</strong> apture Hi-C <strong>O</strong> mnibus <strong>G</strong> ene <strong>S</strong> core and works best with pcHi-C datasets processed using <a href="https://bioconductor.org/packages/release/bioc/html/Chicago.html">CHiCAGO</a>.</p>
</div>
<div id="what-does-rcogs-require" class="section level2">
<h2 class="hasAnchor">
<a href="#what-does-rcogs-require" class="anchor"></a>What does rCOGS require ?</h2>
<p>The hardest part of running rCOGS is making sure source files are available and in the correct format. With all files make sure they are relevant to the population that you are studying and use the correct genome build coordinates. <strong>Unless explicitly stated all coords are GRCh37 and non-zero based.</strong></p>
<p>Full support files that are suitable for use with data from <a href="http://dx.doi.org/10.1016/j.cell.2016.09.037">Javierre et al.</a> are available in <code>./inst/extdata/</code>. Note that this vignette uses cutdown versions of these files filtered to use only chr22 to speed things up.</p>
<p>Depending on usage rCOGS requires the following files:</p>
<div id="approximate-linkage-disequilibrium-independent-region-files" class="section level3">
<h3 class="hasAnchor">
<a href="#approximate-linkage-disequilibrium-independent-region-files" class="anchor"></a>approximate linkage disequilibrium independent region files</h3>
<p>These could be computed using data from the <a href="ftp://ftp.ncbi.nlm.nih.gov/hapmap/recombination/latest/rates/">International HapMap project</a> or you can use more recent alternative regions from <a href="https://doi.org/10.1093/bioinformatics/btv546">Berisa et al.</a>. Note that your LD file should match the population that you are studying! If you are using a more targeted platform such as ImmunoChip then you should use the LD regions used to design the platform. What is key is that these regions are non overlapping, it does not matter if they are not contiguous (you will just lose associations that don’t overlap). The format is non-zero based BED format as follows</p>
<table class="table">
<thead><tr class="header">
<th>chr</th>
<th>start</th>
<th>end</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>1</td>
<td>888659</td>
</tr>
<tr class="even">
<td>1</td>
<td>888660</td>
<td>1891263</td>
</tr>
<tr class="odd">
<td>1</td>
<td>1891264</td>
<td>2299651</td>
</tr>
</tbody>
</table>
</div>
<div id="minor-allele-frequency-estimates-in-controls" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-allele-frequency-estimates-in-controls" class="anchor"></a>Minor allele frequency estimates in controls</h3>
<p>In order to model whether a given SNP is causal rCOGS needs an estimate of the Minor allele Frequency (MAF) in a control set of samples. If you are using your own data you will have this information to hand. If on the other hand you are using summary statistics downloaded from a public resource such as <a href="https://www.ebi.ac.uk/gwas/">GWAS Catalog</a> you can use a reference set of genotypes such as <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">1KGenomes</a> again it is important to make sure that these are relevant to the population that you are studying. The format is as follows and each variant to assessed should have a MAF in this file otherwise it will be excluded. Again position are non-zero based.</p>
<table class="table">
<thead><tr class="header">
<th>chr</th>
<th>pos</th>
<th>maf</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>713914</td>
<td>0.033325</td>
</tr>
<tr class="even">
<td>1</td>
<td>714310</td>
<td>0.026845</td>
</tr>
<tr class="odd">
<td>1</td>
<td>715265</td>
<td>0.037027</td>
</tr>
</tbody>
</table>
</div>
<div id="trait-association-statistics" class="section level3">
<h3 class="hasAnchor">
<a href="#trait-association-statistics" class="anchor"></a>Trait association statistics</h3>
<p>These are the univariate p-values obtained from GWAS analysis, nowadays the fitting of regression models is mainly performed in <a href="https://www.cog-genomics.org/plink2">PLINK</a> and thus this output format is supported. If however you have used an alternative method to assess association then it should have the following format.</p>
<table class="table">
<thead><tr class="header">
<th>chr</th>
<th>pos</th>
<th>p</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>768253</td>
<td>0.617</td>
</tr>
<tr class="even">
<td>1</td>
<td>781845</td>
<td>0.891</td>
</tr>
<tr class="odd">
<td>1</td>
<td>787606</td>
<td>0.871</td>
</tr>
<tr class="even">
<td>1</td>
<td>787844</td>
<td>0.977</td>
</tr>
</tbody>
</table>
</div>
<div id="pchi-c-data" class="section level3">
<h3 class="hasAnchor">
<a href="#pchi-c-data" class="anchor"></a>pcHi-C data</h3>
<p>By far the most complicated data file this should have the following format with data taken from <a href="http://dx.doi.org/10.1016/j.cell.2016.09.037">Javierre et al.</a>. It is an enhanced version of the peakMatrix format output by CHiCAGO.</p>
<table class="table">
<colgroup>
<col width="3%">
<col width="1%">
<col width="3%">
<col width="1%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="2%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="5%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="4%">
<col width="5%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
</colgroup>
<thead><tr class="header">
<th>ensg</th>
<th>name</th>
<th>biotype</th>
<th>strand</th>
<th>baitChr</th>
<th>baitStart</th>
<th>baitEnd</th>
<th>baitID</th>
<th>baitName</th>
<th>oeChr</th>
<th>oeStart</th>
<th>oeEnd</th>
<th>oeID</th>
<th>oeName</th>
<th>dist</th>
<th>Monocytes</th>
<th>Macrophages_M0</th>
<th>Macrophages_M1</th>
<th>Macrophages_M2</th>
<th>Neutrophils</th>
<th>Megakaryocytes</th>
<th>Endothelial_precursors</th>
<th>Erythroblasts</th>
<th>Foetal_thymus</th>
<th>Naive_CD4</th>
<th>Total_CD4_MF</th>
<th>Total_CD4_Activated</th>
<th>Total_CD4_NonActivated</th>
<th>Naive_CD8</th>
<th>Total_CD8</th>
<th>Naive_B</th>
<th>Total_B</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>ENSG00000000003</td>
<td>TSPAN6</td>
<td>protein_coding</td>
<td>-</td>
<td>X</td>
<td>99894205</td>
<td>99902395</td>
<td>813869</td>
<td>SRPX2;TSPAN6</td>
<td>X</td>
<td>100023790</td>
<td>100032508</td>
<td>813903</td>
<td>.</td>
<td>129849</td>
<td>1.2768325567372</td>
<td>3.36925757955278</td>
<td>3.50427556762941</td>
<td>5.77885715736681</td>
<td>0.71946319470011</td>
<td>1.57426075233114</td>
<td>0.935151981213395</td>
<td>2.07176252723988</td>
<td>0.518991005066328</td>
<td>1.35844284079167</td>
<td>1.80051450777455</td>
<td>2.14482233107626</td>
<td>2.81450087489112</td>
<td>1.0646936885784</td>
<td>0.329294941260539</td>
<td>2.31949322698618</td>
<td>2.91843729769059</td>
</tr>
<tr class="even">
<td>ENSG00000000003</td>
<td>TSPAN6</td>
<td>protein_coding</td>
<td>-</td>
<td>X</td>
<td>99894205</td>
<td>99902395</td>
<td>813869</td>
<td>SRPX2;TSPAN6</td>
<td>X</td>
<td>100038149</td>
<td>100039492</td>
<td>813905</td>
<td>.</td>
<td>140521</td>
<td>2.22145436128498</td>
<td>2.34713825641161</td>
<td>1.94777355869935</td>
<td>1.8326996671119</td>
<td>1.19340351282302</td>
<td>4.71363557745529</td>
<td>1.75066400673428</td>
<td>3.10545714323329</td>
<td>1.8506228303492</td>
<td>1.52337854148901</td>
<td>1.92401193124468</td>
<td>1.4553080801738</td>
<td>0.959541835366562</td>
<td>1.47904009071165</td>
<td>2.33301812072822</td>
<td>5.09250926114187</td>
<td>3.5849349597336</td>
</tr>
</tbody>
</table>
</div>
<div id="restriction-fragment-digest-file" class="section level3">
<h3 class="hasAnchor">
<a href="#restriction-fragment-digest-file" class="anchor"></a>Restriction fragment digest file</h3>
<p>This file represents the output of an <em>in</em> <em>silico</em> restriction digest for the restriction enzyme employed in the pcHi-C experiment. At the time that this was written this was <em>Hind</em> III. These regions should be zero based and match those in the pcHi-C file above, additionally, the <strong>fragid</strong> column should refer to the <strong>baitID</strong> and <strong>oeID</strong> columns in the pcHi-C dataset above.</p>
<table class="table">
<thead><tr class="header">
<th>chr</th>
<th>start</th>
<th>end</th>
<th>fragid</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>1</td>
<td>16007</td>
<td>1</td>
</tr>
<tr class="even">
<td>1</td>
<td>16008</td>
<td>24571</td>
<td>2</td>
</tr>
<tr class="odd">
<td>1</td>
<td>24572</td>
<td>27981</td>
<td>3</td>
</tr>
<tr class="even">
<td>1</td>
<td>27982</td>
<td>30429</td>
<td>4</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="pchi-c-designannotation-file" class="section level2">
<h2 class="hasAnchor">
<a href="#pchi-c-designannotation-file" class="anchor"></a>pcHi-C design/annotation file</h2>
<p>This file is a list of all the fragments for which capture probes were designed and depends on what annotation you are using. The <strong>fragid</strong> column again should be compatible with the restriction digest file and pcHi-C file. The <strong>ensg</strong> should also map to the <strong>ensg</strong> file in pcHi-C data file. <strong>Note that a fragment can contain more than one promoter !</strong></p>
<table class="table">
<thead><tr class="header">
<th>fragid</th>
<th>ensg</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>218</td>
<td>ENSG00000272438</td>
</tr>
<tr class="even">
<td>218</td>
<td>ENSG00000230699</td>
</tr>
<tr class="odd">
<td>219</td>
<td>ENSG00000241180</td>
</tr>
<tr class="even">
<td>220</td>
<td>ENSG00000268179</td>
</tr>
<tr class="odd">
<td>220</td>
<td>ENSG00000223764</td>
</tr>
<tr class="even">
<td>220</td>
<td>ENSG00000187634</td>
</tr>
<tr class="odd">
<td>223</td>
<td>ENSG00000187961</td>
</tr>
</tbody>
</table>
<div id="coding-snps" class="section level3">
<h3 class="hasAnchor">
<a href="#coding-snps" class="anchor"></a>Coding SNPs</h3>
<p>rCOGS integrates functional information where possible. To do this it needs a catalogue of exonic variants. This can be obtained from a tool such as <a href="https://doi.org/10.1186/s13059-016-0974-4">VEP</a> and needs to have the following format:</p>
<table class="table">
<thead><tr class="header">
<th>chr</th>
<th>pos</th>
<th>ensg</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>99883719</td>
<td>ENSG00000000003</td>
</tr>
<tr class="even">
<td>1</td>
<td>99883962</td>
<td>ENSG00000000003</td>
</tr>
<tr class="odd">
<td>1</td>
<td>99884017</td>
<td>ENSG00000000003</td>
</tr>
<tr class="even">
<td>1</td>
<td>99884166</td>
<td>ENSG00000000034</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="installing-cogs" class="section level2">
<h2 class="hasAnchor">
<a href="#installing-cogs" class="anchor"></a>Installing COGS</h2>
<p>You will need to have the devtool R package installed</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">devtools</span>)
<span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"ollyburren/rCOGS"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rCOGS</span>)
<span class="co">#&gt; Loading required package: data.table</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">GenomicRanges</span>)
<span class="co">#&gt; Loading required package: stats4</span>
<span class="co">#&gt; Loading required package: BiocGenerics</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'BiocGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:parallel':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
<span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
<span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     IQR, mad, sd, var, xtabs</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
<span class="co">#&gt;     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
<span class="co">#&gt;     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
<span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
<span class="co">#&gt;     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
<span class="co">#&gt;     union, unique, unsplit, which, which.max, which.min</span>
<span class="co">#&gt; Loading required package: S4Vectors</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following objects are masked from 'package:data.table':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     first, second</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid</span>
<span class="co">#&gt; Loading required package: IRanges</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'IRanges'</span>
<span class="co">#&gt; The following object is masked from 'package:data.table':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     shift</span>
<span class="co">#&gt; Loading required package: GenomeInfoDb</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">knitr</span>)</pre></body></html></div>
</div>
<div id="loading-recombination-data" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-recombination-data" class="anchor"></a>loading recombination data</h2>
<p>Here we load in ld regions - these are checked and output as a <strong>GRanges</strong> object</p>
<p>This is how to build objects - note due to size you will need to download these supporting objects into ./inst/extdata/</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">ld.gr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/load_ld_regions.html">load_ld_regions</a></span>(<span class="st">'../inst/extdata/hamap_1cM_recomb.bed'</span>)
<span class="no">ld.gr</span></pre></body></html></div>
<pre><code>#&gt; GRanges object with 69 ranges and 1 metadata column:
#&gt;        seqnames             ranges strand |      ldid
#&gt;           &lt;Rle&gt;          &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
#&gt;    [1]       22         1-16252600      * |      3385
#&gt;    [2]       22  16252601-17089639      * |      3386
#&gt;    [3]       22  17089640-17571604      * |      3387
#&gt;    [4]       22  17571605-17718606      * |      3388
#&gt;    [5]       22  17718607-17823261      * |      3389
#&gt;    ...      ...                ...    ... .       ...
#&gt;   [65]       22  49349502-49592082      * |      3449
#&gt;   [66]       22  49592083-49777002      * |      3450
#&gt;   [67]       22  49777003-49910517      * |      3451
#&gt;   [68]       22  49910518-50835040      * |      3452
#&gt;   [69]       22 50835041-536870911      * |      3453
#&gt;   -------
#&gt;   seqinfo: 22 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="loading-maf-data-filter-at-5" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-maf-data-filter-at-5" class="anchor"></a>loading MAF data filter at 5%</h2>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">maf.DT</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/load_ref_maf.html">load_ref_maf</a></span>(<span class="st">'../inst/extdata/uk10k_reference_maf.tab'</span>,<span class="kw">min.maf</span><span class="kw">=</span><span class="fl">0.05</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">maf.DT</span>)</pre></body></html></div>
<pre><code>#&gt;    chr      pos      maf         pid
#&gt; 1:  22 16051249 0.107115 22:16051249
#&gt; 2:  22 16053444 0.072997 22:16053444
#&gt; 3:  22 16054667 0.244909 22:16054667
#&gt; 4:  22 16055122 0.085559 22:16055122
#&gt; 5:  22 16057417 0.107908 22:16057417
#&gt; 6:  22 16490618 0.342238 22:16490618</code></pre>
</div>
<div id="load-association-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-association-data" class="anchor"></a>load association data</h2>
<p>N.B here we have a study with 4036 cases and 6959 controls.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">t1d.DT</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/load_gwas.html">load_gwas</a></span>(<span class="st">'../inst/extdata/ncooper_t1d_gwas.tab'</span>,<span class="no">maf.DT</span>,<span class="no">ld.gr</span>,<span class="kw">n.cases</span><span class="kw">=</span><span class="fl">5913</span>,<span class="kw">n.controls</span><span class="kw">=</span><span class="fl">8829</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">t1d.DT</span>)</pre></body></html></div>
<pre><code>#&gt;            pid chr      pos     p      maf ld          ppi
#&gt; 1: 22:16849941  22 16849941 0.798 0.449220  2 1.221142e-05
#&gt; 2: 22:16851640  22 16851640 0.419 0.374769  2 1.675352e-05
#&gt; 3: 22:16851673  22 16851673 0.419 0.374769  2 1.675352e-05
#&gt; 4: 22:16851899  22 16851899 0.419 0.374372  2 1.675700e-05
#&gt; 5: 22:16852303  22 16852303 0.395 0.368024  2 1.740740e-05
#&gt; 6: 22:16852914  22 16852914 0.411 0.368553  2 1.699918e-05</code></pre>
<p>Notice that not all variants are imported, some are ommited as they are below the MAF filter (5%) used in the previous step others don’t overlap an LD regions.</p>
</div>
<div id="load-in-pchi-c-map" class="section level2">
<h2 class="hasAnchor">
<a href="#load-in-pchi-c-map" class="anchor"></a>load in pcHi-C map</h2>
<p>note the biotype.filter argument. This can be a vector of biotypes or left blank if everything is required.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">feature.sets</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/make_pchic.html">make_pchic</a></span>(<span class="st">'../inst/extdata/javierre_pchic_ensembl75.tab'</span>,<span class="kw">biotype.filter</span><span class="kw">=</span><span class="st">'protein_coding'</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">feature.sets</span>)</pre></body></html></div>
<pre><code>#&gt;  [1] "Monocytes"              "Macrophages_M0"         "Macrophages_M1"        
#&gt;  [4] "Macrophages_M2"         "Neutrophils"            "Megakaryocytes"        
#&gt;  [7] "Endothelial_precursors" "Erythroblasts"          "Foetal_thymus"         
#&gt; [10] "Naive_CD4"              "Total_CD4_MF"           "Total_CD4_Activated"   
#&gt; [13] "Total_CD4_NonActivated" "Naive_CD8"              "Total_CD8"             
#&gt; [16] "Naive_B"                "Total_B"                "VProm"</code></pre>
<p>Note that we obtain a list where each element is a <strong>data.table</strong></p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">feature.sets</span><span class="kw">[[</span><span class="fl">1</span>]])
<span class="co">#&gt;               ensg fragid chic.score</span>
<span class="co">#&gt; 1: ENSG00000015475 429638   5.683019</span>
<span class="co">#&gt; 2: ENSG00000015475 429639   6.259106</span>
<span class="co">#&gt; 3: ENSG00000015475 429641   5.181800</span>
<span class="co">#&gt; 4: ENSG00000015475 429651   5.261521</span>
<span class="co">#&gt; 5: ENSG00000015475 429663   5.580373</span>
<span class="co">#&gt; 6: ENSG00000015475 429743   5.604308</span></pre></body></html></div>
</div>
<div id="create-virtual-promoter-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#create-virtual-promoter-regions" class="anchor"></a>create Virtual Promoter regions</h2>
<p>All Hi-C technologies struggle when assessing interactions between adjacent restricition fragments as it is impossible to resolve what is due to stochastic Brownian motion and which are short range interactions. To offset this COGS has a routine that creates a `Virtual’ promter region that consists of a baited fragment and the adjacent 5’ and 3’ fragments. It takes the digest and design file as arguments.</p>
<p>We add this to the list of regions.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co">## get a list of all genes included</span>
<span class="no">all.genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">feature.sets</span>,<span class="kw">function</span>(<span class="no">g</span>) <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">g</span>$<span class="no">ensg</span>)) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="st">'c'</span>,<span class="no">.</span>) <span class="kw">%&gt;%</span> <span class="no">unique</span>
<span class="no">feature.sets</span><span class="kw">[[</span><span class="st">'VProm'</span>]] <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/make_vprom.html">make_vprom</a></span>(<span class="st">'../inst/extdata/hindIII_digest_grch37.tab'</span>,<span class="st">'./inst/extdata/javierre_design_ensembl75.tab'</span>,<span class="no">all.genes</span>)</pre></body></html></div>
</div>
<div id="load-in-coding-snps" class="section level2">
<h2 class="hasAnchor">
<a href="#load-in-coding-snps" class="anchor"></a>load in coding SNPS</h2>
<p>returns a genomic ranges object</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">csnps.gr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/make_csnps.html">make_csnps</a></span>(<span class="st">'../inst/extdata/coding_snps_vep_ensembl75.tab'</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">csnps.gr</span>)</pre></body></html></div>
<pre><code>#&gt; GRanges object with 6 ranges and 2 metadata columns:
#&gt;       seqnames    ranges strand |    fragid            ensg
#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;     &lt;character&gt;
#&gt;   [1]       22  17264320      * |         0 ENSG00000172967
#&gt;   [2]       22  17264341      * |         0 ENSG00000172967
#&gt;   [3]       22  17264369      * |         0 ENSG00000172967
#&gt;   [4]       22  17264400      * |         0 ENSG00000172967
#&gt;   [5]       22  17264498      * |         0 ENSG00000172967
#&gt;   [6]       22  17264565      * |         0 ENSG00000172967
#&gt;   -------
#&gt;   seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="load-digest-file" class="section level2">
<h2 class="hasAnchor">
<a href="#load-digest-file" class="anchor"></a>load Digest file</h2>
<p>The fundamental unit of a pcHi-C map is a restriction fragment this saves us a lot of computation as we can precompute the support for each fragment to be causal once and then combine to get scores.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">digest.gr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/load_digest.html">load_digest</a></span>(<span class="st">'../inst/extdata/hindIII_digest_grch37.tab'</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">digest.gr</span>)</pre></body></html></div>
<pre><code>#&gt; GRanges object with 6 ranges and 1 metadata column:
#&gt;       seqnames            ranges strand |    fragid
#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
#&gt;   [1]       22        1-16058012      * |    429254
#&gt;   [2]       22 16058013-16058591      * |    429255
#&gt;   [3]       22 16058592-16060021      * |    429256
#&gt;   [4]       22 16060022-16067173      * |    429257
#&gt;   [5]       22 16067174-16071272      * |    429258
#&gt;   [6]       22 16071273-16073271      * |    429259
#&gt;   -------
#&gt;   seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="optionalremove-mhc" class="section level2">
<h2 class="hasAnchor">
<a href="#optionalremove-mhc" class="anchor"></a>Optional(Remove MHC)</h2>
<p>MHC/HLA has an extremely complex and long LD structure and the genetic finemapping method employed is unable to effectively deal with it, thus it is best to exclude and analyse separately.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">t1d.DT</span>)
<span class="no">mhc.idx</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">t1d.DT</span>$<span class="no">chr</span><span class="kw">==</span><span class="fl">6</span> <span class="kw">&amp;</span> <span class="fu">between</span>(<span class="no">t1d.DT</span>$<span class="no">pos</span>,<span class="fl">25e6</span>,<span class="fl">35e6</span>))
<span class="kw">if</span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">mhc.idx</span>)<span class="kw">&gt;</span><span class="fl">0</span>)
    <span class="no">t1d.DT</span> <span class="kw">&lt;-</span> <span class="no">t1d.DT</span>[-<span class="no">mhc.idx</span>,]
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">t1d.DT</span>)</pre></body></html></div>
</div>
<div id="compute-overall-cogs-score" class="section level2">
<h2 class="hasAnchor">
<a href="#compute-overall-cogs-score" class="anchor"></a>compute overall cogs score</h2>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">overall.scores</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/compute_cogs.html">compute_cogs</a></span>(<span class="no">t1d.DT</span>,<span class="no">csnps.gr</span>,<span class="no">digest.gr</span>,<span class="no">feature.sets</span>)
<span class="co">#&gt; rCOGS:compute_cogs feature.names argument missing computing overall score using Monocytes,</span>
<span class="co">#&gt; Macrophages_M0,</span>
<span class="co">#&gt; Macrophages_M1,</span>
<span class="co">#&gt; Macrophages_M2,</span>
<span class="co">#&gt; Neutrophils,</span>
<span class="co">#&gt; Megakaryocytes,</span>
<span class="co">#&gt; Endothelial_precursors,</span>
<span class="co">#&gt; Erythroblasts,</span>
<span class="co">#&gt; Foetal_thymus,</span>
<span class="co">#&gt; Naive_CD4,</span>
<span class="co">#&gt; Total_CD4_MF,</span>
<span class="co">#&gt; Total_CD4_Activated,</span>
<span class="co">#&gt; Total_CD4_NonActivated,</span>
<span class="co">#&gt; Naive_CD8,</span>
<span class="co">#&gt; Total_CD8,</span>
<span class="co">#&gt; Naive_B,</span>
<span class="co">#&gt; Total_B,</span>
<span class="co">#&gt; VProm,</span>
<span class="co">#&gt; coding_snp</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">overall.scores</span>[<span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="no">cogs</span>,<span class="kw">decreasing</span> <span class="kw">=</span> <span class="fl">TRUE</span>),]) <span class="kw">%&gt;%</span> <span class="no">kable</span></pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="left">ensg</th>
<th align="right">cogs</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ENSG00000187045</td>
<td align="right">0.9569800</td>
</tr>
<tr class="even">
<td align="left">ENSG00000128342</td>
<td align="right">0.6052610</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000100385</td>
<td align="right">0.2368662</td>
</tr>
<tr class="even">
<td align="left">ENSG00000185133</td>
<td align="right">0.2196913</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000198832</td>
<td align="right">0.2192132</td>
</tr>
<tr class="even">
<td align="left">ENSG00000099954</td>
<td align="right">0.2183529</td>
</tr>
</tbody>
</table>
</div>
<div id="compute-a-t-cell-specific-score" class="section level2">
<h2 class="hasAnchor">
<a href="#compute-a-t-cell-specific-score" class="anchor"></a>Compute a T cell specific score</h2>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">target_tissue</span><span class="kw">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Total_CD4_Activated'</span>,<span class="st">'Total_CD4_NonActivated'</span>,<span class="st">'Naive_CD4'</span>,<span class="st">'Total_CD4_MF'</span>,<span class="st">'Naive_CD8'</span>,<span class="st">'Total_CD8'</span>)
<span class="no">tcell.scores</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/compute_cogs.html">compute_cogs</a></span>(<span class="no">t1d.DT</span>,<span class="no">csnps.gr</span>,<span class="no">digest.gr</span>,<span class="no">feature.sets</span>,<span class="no">target_tissue</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">.</span>,)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">tcell.scores</span>[<span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="no">cogs</span>,<span class="kw">decreasing</span> <span class="kw">=</span> <span class="fl">TRUE</span>),]) <span class="kw">%&gt;%</span> <span class="no">kable</span></pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="left">ensg</th>
<th align="right">cogs</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ENSG00000099954</td>
<td align="right">0.2106387</td>
</tr>
<tr class="even">
<td align="left">ENSG00000099968</td>
<td align="right">0.0401019</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000131100</td>
<td align="right">0.0401019</td>
</tr>
<tr class="even">
<td align="left">ENSG00000069998</td>
<td align="right">0.0056926</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000172967</td>
<td align="right">0.0007576</td>
</tr>
<tr class="even">
<td align="left">ENSG00000182902</td>
<td align="right">0.0000110</td>
</tr>
</tbody>
</table>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Oliver Burren.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
